cluster/kubernetes/update.go

```diff
@@ -81,89 +81,95 @@ func updatePodController(def []byte, newImageID flux.ImageID) ([]byte, error) {
+	if manifest.Metadata.Name == "" {
+		return fmt.Errorf("could not find resource name")
+	}
 
-	if oldImage.Repository() != newImage.Repository() {
-		return fmt.Errorf(`expected existing image name and new image name to match, but %q != %q`, oldImage.Repository(), newImage.Repository())
+	// Check if any containers need updating. As we go through, we calculate the
+	// new manifest name, in case it includes the image tag (as in replication
+	// controllers).
+	newDefName := manifest.Metadata.Name
+	matchingContainers := map[int]Container{}
+	for i, c := range manifest.Spec.Template.Spec.Containers {
+		currentImage, err := flux.ParseImageID(c.Image)
+		if err != nil {
+			return fmt.Errorf("could not parse image %s", c.Image)
+		}
+		if currentImage.Repository() == newImage.Repository() {
+			matchingContainers[i] = c
+		}
+		_, _, oldImageTag := currentImage.Components()
+		if strings.HasSuffix(manifest.Metadata.Name, oldImageTag) {
+			newDefName = manifest.Metadata.Name[:len(manifest.Metadata.Name)-len(oldImageTag)] + newImage.Tag
+		}
 	}
 
-	// Now to replace bits. Specifically,
-	// * the name, with a re-tagged name
-	// * the image for the container
-	// * the version label (in two places)
-	//
 	// Some values (most likely the version) will be interpreted as a
 	// number if unquoted; while, on the other hand, it is apparently
 	// not OK to quote things that don't look like numbers. So: we
 	// extract values *without* quotes, and add them if necessary.
-
-	newDefName := oldDefName
-	_, _, oldImageTag := oldImage.Components()
-	_, _, newImageTag := newImage.Components()
-	if strings.HasSuffix(oldDefName, oldImageTag) {
-		newDefName = oldDefName[:len(oldDefName)-len(oldImageTag)] + newImageTag
-	}
-
 	newDefName = maybeQuote(newDefName)
-	newTag := maybeQuote(newImageTag)
 
-	// fmt.Fprintln(trace, "")
-	// fmt.Fprintln(trace, "Replacing ...")
-	// fmt.Fprintf(trace, "Resource name: %s -> %s\n", oldDefName, newDefName)
-	// fmt.Fprintf(trace, "Version in templates (and selector if present): %s -> %s\n", oldImageTag, newTag)
-	// fmt.Fprintf(trace, "Image in templates: %s -> %s\n", oldImage, newImage)
-	// fmt.Fprintln(trace, "")
+	if len(matchingContainers) == 0 {
+		return fmt.Errorf("could not find container using image: %s", newImage.Repository())
+	}
 
-	// The name we want is that under `metadata:`, which will be indented once
-	replaceRCNameRE := regexp.MustCompile(`(?m:^(  name:\s*) (?:"?[\w-]+"?)(\s.*)$)`)
-	withNewDefName := replaceRCNameRE.ReplaceAllString(def, fmt.Sprintf(`$1 %s$2`, newDefName))
+	// Detect how indented the "containers" block is.
+	newDef := string(def)
+	matches := regexp.MustCompile(`( +)containers:.*`).FindStringSubmatch(newDef)
+	if len(matches) != 2 {
+		return fmt.Errorf("could not find container specs")
+	}
+	indent := matches[1]
+
+	// Replace the container images
+	containersRE := regexp.MustCompile(`(?m:^` + indent + `containers:\s*(?:#.*)*$(?:\n(?:` + indent + `[-\s].*)?)*)`)
```

Comment:
Some comments saying what these regexps are for (and more distinct names) would be good here.

(Yes, I am the original offender .. I have sought forgiveness at the altar of Parsec and will continue to do so til my dying day)